archethic-foundation/archethic-node
# AEIP-19 token recipients

# Description

/!\ Base branch 1174-token-resupply (AEIP-18) 
This is AEIP-19 implementation. This adds a `recipients` field in the token transactions. Minted tokens can be sent to recipients in the same transaction.

Fixes #1175 

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Scenario 1: Create a token with recipients
```yaml
transaction_type: token
content: |
  {
    "supply": 10000000000,
    "type": "fungible",
    "decimals": 8,
    "name": "COINCOIN",
    "allow_mint": true,
    "recipients": [
      {"to": "00002223bbd4ec3d64ae597696c7d7ade1cee65c639d885450ad2d7b75592ac76afa", "amount": 333}
    ],
    "aeip": [2, 18]
  }
```
```
go run main.go send-transaction --config newtoken.yaml --access-seed bastien --index 0
```
- Check the movements in resulting

Scenario 2: 
- Create a resupply with recipients
```yaml
transaction_type: token
content: |
  {
    "aeip": [2,18],
    "supply": 100000000,
    "recipients": [{"to": "00002223bbd4ec3d64ae597696c7d7ade1cee65c639d885450ad2d7b75592ac76afa", "amount": 100000000}],
    "token_reference": "000058C44C933FBD047BF694490371BB41DA037597FFD57970A1DE6832AF990B7752"
  }
```
```
go run main.go send-transaction --config supply.yaml --access-seed bastien --index 1
```
- Check the movements in resulting


Many unit tests

# Checklist:

- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
- Any dependent changes have been merged and published in downstream modules
